# INFO 371 - Comparing Models

For this homework assignment, we are going to ask to you to put together everything you have learned so far to analyze 4 very different datasets! The goal is for you to gain experience and intuition for how each of the supervised learning algorithms perform in different scenarios. 

# Datasets 
There are four datasets you will need to analyze (one of which you have worked with in past asignments!) 

## Dataset 1: Soybean Dataset
With this dataset (``soybean.csv``) your goal is to predict which of the four diseases a particular soy-bean plant has. This datset was part of a project used to build a survey to diagnose crops (you can read more about this work [here](http://www.mli.gmu.edu/papers/79-80/80-2.pdf). The dataset has the following features: 

* Month: which month was the plant disease found, represented as a number meaning april is 4 ect.
* Plant Stand: describes how much of the soybean plant is above the soil 
    * 0 is avg,  1 is below avg
* Precipitation:  0: below avg, 1: avg, 2: above avg
* temp:  0: below avg, 1: avg, 2: above avg
* hail: 0 is no, 1 is yes 
* crop-hist	
    * 0: diff-lst-year
    * 1: same-lst-yr
    * 2: same-lst-two-yrs
    * 3: same-lst-sev-yrs
* area-damaged
    * 0: scattered
    * 1: low-areas
    * 2: upper-areas
    * 3:whole-field
* severity
    * 0 minor,1: potentially severe, 2: severe
* seed-tmt: 0: None, 1: used fungicide 
* germination: How much of the plants sprouted
    * 0: 90-100%, 1: 80-89%, 2: 80% or less	
* plant-growth: 0: normal, 1: abnormal 
* leaves: 0: normal, 1: abnormal 
* leafspots-halo: 0: absent: 1: has yellow halos, 2: has halos, but not yellow halos
* leafspots-margin: 0: water-soaked margin,1: no margin, 2: does not apply 	
* leafspot-size: 0: less than 1/8 in, 1: bigger than 1/8 in, 2: does not apply
* leaf-shread: 	0: absent, 1: present
* leaf-malf: 0: absent, 1: present
* leaf-mild: 0: absent, 1: present on upper surface ,2: present on lower surface
* stem: 0: normal, 1: not normal 
* lodging: 0: absent, 1: present
* stem-cankers: 0: absent, 1: found below soil, 2: found above soil, 3: found above second node
* canker-lesion	0: absent,1: brown color ,2: dark brown to black color, 3: tan
* fruiting-bodies: 	0: absent, 1: present
* extern_decay: 0: absent, 1: firm and dry, 2: watery 
* mycelium: 0: absent, 1: present	
* int-discolor: 0: none, 1: black, 2: brown 
* sclerotia: 0: absent, 1: present	
* fruit-pods: 0: normal, 1: diseased, 2: few present, 3: does not apply 	
* fruit-spots: 0: absent, 1: colored, 2: brown with black specks, 3: distorted, 4: does not apply	
* seed: 0: normal, 1: not normal
* mold-groth: 0: absent, 1: present
* seed-discolor: 0: absent, 1: present
* seed-size: 0: average, 1: below average 	
* shriveling: 0: absent, 1: present	
* roots	: 0: normal, 1: diseased 


* Label: which of the diseases that plat has
    * D1: diaporthe stem canker
    * D2: charcoal rot
    * D3: rhizoctonia root rot
    * D4: phytophthora rot


## Dataset 2: Iris Dataset
With this dataset (``iris.csv``), your goal to categorize Iris flowers based on the four measurements (aka: features) into the correct species (aka: labels) setosa, virginica, and versicolor. The data contains 50 flowers of each species (150 in total), and four measurements for each species (petal length and width, and sepal length and width). All of these are numeric measures.

## Dataset 3: Skin Segmentation Dataset
With this dataset (``Skin_NonSkin.tsv``), your goal is to predict whether a particular pixel is a skin tone or not a skin tone. Each row in the dataset is a pixel taken from a random image. The features are the values in the B, G, R color space (i.e blue, green, red). 

## Dataset 4: Taiwanese Bankruptcy Prediction Dataset 
With this dataset (``bankruptcy.csv``), your goal is to predict whether a company went backrupt or not. This dataset was collected from the Taiwan Economic Journal for the years 1999 to 2009. Company bankruptcy was defined based on the business regulations of the Taiwan Stock Exchange. The dataset has the following features: 

* Bankrupt?: Class label
* ROA(C) before interest and depreciation before interest: Return On Total Assets(C)
* ROA(A) before interest and % after tax: Return On Total Assets(A)
* ROA(B) before interest and depreciation after tax: Return On Total Assets(B)
* Operating Gross Margin: Gross Profit/Net Sales
* Realized Sales Gross Margin: Realized Gross Profit/Net Sales
* Operating Profit Rate: Operating Income/Net Sales
* Pre-tax net Interest Rate: Pre-Tax Income/Net Sales
* After-tax net Interest Rate: Net Income/Net Sales
* Non-industry income and expenditure/revenue: Net Non-operating Income Ratio
* Continuous interest rate (after tax): Net Income-Exclude Disposal Gain or Loss/Net Sales
* Operating Expense Rate: Operating Expenses/Net Sales
* Research and development expense rate: (Research and Development Expenses)/Net Sales
* Cash flow rate: Cash Flow from Operating/Current Liabilities
* Interest-bearing debt interest rate: Interest-bearing Debt/Equity
* Tax rate (A): Effective Tax Rate
* Net Value Per Share (B): Book Value Per Share(B)
* Net Value Per Share (A): Book Value Per Share(A)
* Net Value Per Share (C): Book Value Per Share(C)
* Persistent EPS in the Last Four Seasons: EPS-Net Income
* Cash Flow Per Share
* Revenue Per Share (Yuan ¥): Sales Per Share
* Operating Profit Per Share (Yuan ¥): Operating Income Per Share
* Per Share Net profit before tax (Yuan ¥): Pretax Income Per Share
* Realized Sales Gross Profit Growth Rate
* Operating Profit Growth Rate: Operating Income Growth
* After-tax Net Profit Growth Rate: Net Income Growth
* Regular Net Profit Growth Rate: Continuing Operating Income after Tax Growth
* Continuous Net Profit Growth Rate: Net Income-Excluding Disposal Gain or Loss Growth
* Total Asset Growth Rate: Total Asset Growth
* Net Value Growth Rate: Total Equity Growth
* Total Asset Return Growth Rate Ratio: Return on Total Asset Growth
* Cash Reinvestment %: Cash Reinvestment Ratio
* Current Ratio
* Quick Ratio: Acid Test
* Interest Expense Ratio: Interest Expenses/Total Revenue
* Total debt/Total net worth: Total Liability/Equity Ratio
* Debt ratio %: Liability/Total Assets
* Net worth/Assets: Equity/Total Assets
* Long-term fund suitability ratio (A): (Long-term Liability+Equity)/Fixed Assets
* Borrowing dependency: Cost of Interest-bearing Debt
* Contingent liabilities/Net worth: Contingent Liability/Equity
* Operating profit/Paid-in capital: Operating Income/Capital
* Net profit before tax/Paid-in capital: Pretax Income/Capital
* Inventory and accounts receivable/Net value: (Inventory+Accounts Receivables)/Equity
* Total Asset Turnover
* Accounts Receivable Turnover
* Average Collection Days: Days Receivable Outstanding
* Inventory Turnover Rate (times)
* Fixed Assets Turnover Frequency
* Net Worth Turnover Rate (times): Equity Turnover
* Revenue per person: Sales Per Employee
* Operating profit per person: Operation Income Per Employee
* Allocation rate per person: Fixed Assets Per Employee
* Working Capital to Total Assets
* Quick Assets/Total Assets
* Current Assets/Total Assets
* Cash/Total Assets
* Quick Assets/Current Liability
* Cash/Current Liability
* Current Liability to Assets
* Operating Funds to Liability
* Inventory/Working Capital
* Inventory/Current Liability
* Current Liabilities/Liability
* Working Capital/Equity
* Current Liabilities/Equity
* Long-term Liability to Current Assets
* Retained Earnings to Total Assets
* Total income/Total expense
* Total expense/Assets
* Current Asset Turnover Rate: Current Assets to Sales
* Quick Asset Turnover Rate: Quick Assets to Sales
* Working capitcal Turnover Rate: Working Capital to Sales
* Cash Turnover Rate: Cash to Sales
* Cash Flow to Sales
* Fixed Assets to Assets
* Current Liability to Liability
* Current Liability to Equity
* Equity to Long-term Liability
* Cash Flow to Total Assets
* Cash Flow to Liability
* CFO to Assets
* Cash Flow to Equity
* Current Liability to Current Assets
* Liability-Assets Flag: 1 if Total Liability exceeds Total Assets, 0 otherwise
* Net Income to Total Assets
* Total assets to GNP price
* No-credit Interval
* Gross Profit to Sales
* Net Income to Stockholder's Equity
* Liability to Equity
* Degree of Financial Leverage (DFL)
* Interest Coverage Ratio (Interest expense to EBIT)
* Net Income Flag: 1 if Net Income is Negative for the last two years, 0 otherwise
* Equity to Liability


---
# Dataset Exploration 

1. Load in each dataset, then report on how many rows and columns each dataset has AND the label distibution (i.e how many data items are in each label -- for example, the iris dataset has 50 flowers across each of the 3 flower species labels)


2. Based on this dataset size information and the given dataset decriptions alone, which of the supervised learning models (Logistic Regression, K-NN, Naive Bayes, SVMs, decision trees, or random forests) do you think will peform best on each dataset? Which do you think will perform worst? Explain your reasoning for each choice. 
    * NOTE: DO NOT RUN ANY EXPERIMENTS BEFOREHAND! You should use your intuition and class discussions for this! 
    

3. Get a baseline accuracy using the naive/constant model (i.e. a model where you assign the same label to all your testing data and that label is the one that appeared the most in your training data) for each of the datasets. 

In [1]:
import time

import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

#code goes here
soy = pd.read_csv('soybean.csv')
iris = pd.read_csv('iris.csv', delimiter="\t")
skin = pd.read_csv('Skin_NonSkin.tsv', delimiter='\t')
bank = pd.read_csv('bankruptcy.csv')

print("soybean dataset rows, columns: ", soy.shape)
print("iris dataset rows, columns: ", iris.shape)
print("skin dataset rows, columns: ", skin.shape)
print("bank dataset rows, columns: ", bank.shape)
print()
print("soybean dataset label quantities: ", soy.Label.value_counts())
print()
print("iris dataset label quantities: ", iris.Species.value_counts())
print()
print("skin dataset label quantities: ", skin.Label.value_counts())
print()
print("bank dataset label quantities: ", bank[["Bankrupt?"]].value_counts())

soybean dataset rows, columns:  (47, 36)
iris dataset rows, columns:  (150, 5)
skin dataset rows, columns:  (245057, 4)
bank dataset rows, columns:  (6819, 96)

soybean dataset label quantities:  Label
D4    17
D1    10
D2    10
D3    10
Name: count, dtype: int64

iris dataset label quantities:  Species
setosa        50
versicolor    50
virginica     50
Name: count, dtype: int64

skin dataset label quantities:  Label
2    194198
1     50859
Name: count, dtype: int64

bank dataset label quantities:  Bankrupt?
0            6599
1             220
Name: count, dtype: int64


Based on the results of the analysis I believe the skin dataset will perform best with KNN as it is quite large. I believe the iris dataset will work best with SVM given its size and simplicity. I believe the soybean dataset may work well with a tree structure given its smaller size, perhaps logistic regression too. Lastly I believe the banking dataset will work best with naive bayes given its width

In [2]:
from sklearn.metrics import accuracy_score
import numpy as np
#NAIVE MODELS
from sklearn.model_selection import train_test_split
x1 = soy.drop(columns='Label')
y1 = soy['Label']
x1_train, x1_test, y1_train, y1_test = train_test_split(x1, y1, test_size=0.2)

x2 = iris.drop(columns='Species')
y2 = iris['Species']
x2_train, x2_test, y2_train, y2_test = train_test_split(x2, y2, test_size=0.2)

x3 = skin.drop(columns='Label')
y3 = skin['Label']
x3_train, x3_test, y3_train, y3_test = train_test_split(x3, y3, test_size=0.2)

x4 = bank.drop(columns='Bankrupt?')
y4 = bank['Bankrupt?']
x4_train, x4_test, y4_train, y4_test = train_test_split(x4, y4, test_size=0.2)

naive_soy = np.full((len(y1_test)), "D4")
naive_iris = np.full((len(y2_test)), "Setosa")
naive_skin = np.full((len(y3_test)), 2)
naive_bank = np.full((len(y4_test)), 0)

print("Naive soy model accuracy: ",accuracy_score(y1_test, naive_soy))
print("Naive iris model accuracy: ",accuracy_score(y2_test, naive_iris))
print("Naive skin model accuracy: ",accuracy_score(y3_test, naive_skin))
print("Naive bank model accuracy: ",accuracy_score(y4_test, naive_bank))

Naive soy model accuracy:  0.3
Naive iris model accuracy:  0.0
Naive skin model accuracy:  0.7939076144617645
Naive bank model accuracy:  0.9604105571847508


# Model Comparisons and Analysis

4.Compare the average testing accuracy as well as runtime across each of the 4 datasets for each of the following experiment set-ups: 

    1. Logistic Regression 

    2. K-Nearest Neighbors (K-NN) for euclidean and manhattan distance and for k values = {5, 10, 15, 20, 25, 30, 35, 40, 45, 50}

    3. Naive Bayes 

    4. Decision Trees using citerion = "entropy" across various values for max_depth  
    
    5. Random Forests using citerion = "entropy" across various values for num_estimators (try to pick atleast 3 numbers: one using the "ideal" value we talked about in class, one less than the ideal, and one much greater than the ideal) 
    
    6. SVMs using both the Linear and RBF kernel
    

Note: you may use k-fold cross validation for these experiments


5. Were your predicitons from step 2 correct? Why or why not? 


6. Which model would you chose for each dataset? Explain your reasoning and use the results of your experiments and any domain knowledge you have to support your argument. Make sure you discuss in detail the trade-offs you considered as part of this analysis. Your response here should be pretty detailed. 


7. Why do you think certain models performed better on some of your datasets versus others? Are there certain attributes/aspects of each dataset that some machine learning models prefer? Explain your reasoning here in detail. 



In [3]:
# code goes here 

#Logistic Regression
t0 = time.time()
model1 = LogisticRegression()
model1.fit(x1_train, y1_train)
print('logistic soy performance: ', model1.score(x1_test, y1_test))
t1 = time.time()
total_time1 = t1 - t0
print('logistic soy speed: ',total_time1)
print()
t0 = time.time()
model2 = LogisticRegression(max_iter=1000)
model2.fit(x2_train, y2_train)
print('logistic iris performance: ', model2.score(x2_test, y2_test))
t1 = time.time()
total_time2 = t1 - t0
print('logistic iris speed: ',total_time2)
print()
t0 = time.time()
model3 = LogisticRegression()
model3.fit(x3_train, y3_train)
print('logistic skin performance: ', model3.score(x3_test, y3_test))
t1 = time.time()
total_time3 = t1 - t0
print('logistic skin speed: ',total_time3)
print()
t0 = time.time()
model4 = LogisticRegression(max_iter=1000)
model4.fit(x4_train, y4_train)
print('logistic bank performance: ', model4.score(x4_test, y4_test))
t1 = time.time()
total_time4 = t1 - t0
print('logistic bank speed: ',total_time4)
print()

logistic soy performance:  1.0
logistic soy speed:  0.015298843383789062

logistic iris performance:  1.0
logistic iris speed:  0.01629781723022461

logistic skin performance:  0.9200195870399086
logistic skin speed:  1.3564426898956299

logistic bank performance:  0.9530791788856305
logistic bank speed:  1.699185848236084



In [4]:
#KNN EUCLIDEAN
#model1_accuracy = []
model2_accuracy = []
model3_accuracy = []
model4_accuracy = []
model2_time = []
model3_time = []
model4_time = []
for k in range(5,51, 5):
        
###NOTE THE SOY DATASET IS TOO SMALL TO USE FOR KNN GIVEN K > 40 does not work with training set size of 37.
        
        #model1 = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
        #model1.fit(x1_train, y1_train)
        #model1_accuracy.append(model1.score(x1_test, y1_test))
        
        t0 = time.time()
        model2 = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
        model2.fit(x2_train, y2_train)
        model2_accuracy.append(model2.score(x2_test, y2_test))
        t1 = time.time()
        total_time2 = t1 - t0
        model2_time.append(total_time2)
        
            
        t0 = time.time()
        model3 = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
        model3.fit(x3_train, y3_train)
        model3_accuracy.append(model3.score(x3_test, y3_test))
        t1 = time.time()
        total_time3 = t1 - t0
        model3_time.append(total_time3)
        
           
        t0 = time.time()
        model4 = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
        model4.fit(x4_train, y4_train)
        model4_accuracy.append(model4.score(x4_test, y4_test))
        t1 = time.time()
        total_time4 = t1 - t0
        model4_time.append(total_time4)

In [5]:
from numpy import mean
print("KNN EUCLIDEAN MODELLING")
print("Model iris mean Accuracies:", mean(model2_accuracy))
print("Model skin mean Accuracies:", mean(model3_accuracy))
print("Model bank mean Accuracies:", mean(model4_accuracy))
print("Model iris mean Times:", mean(model2_time))
print("Model skin mean Times:", mean(model3_time))
print("Model bank mean Times:", mean(model4_time))

KNN EUCLIDEAN MODELLING
Model iris mean Accuracies: 0.95
Model skin mean Accuracies: 0.9989104709050844
Model bank mean Accuracies: 0.9599706744868037
Model iris mean Times: 0.006351351737976074
Model skin mean Times: 2.9919050216674803
Model bank mean Times: 0.11887910366058349


In [6]:
#KNN manhattan
#model1_accuracy = []
model2_accuracy = []
model3_accuracy = []
model4_accuracy = []
model2_time = []
model3_time = []
model4_time = []
for k in range(5,51, 5):
        
###NOTE THE SOY DATASET IS TOO SMALL TO USE FOR KNN GIVEN K > 40 does not work with training set size of 37.
        
        #model1 = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
        #model1.fit(x1_train, y1_train)
        #model1_accuracy.append(model1.score(x1_test, y1_test))
        
        t0 = time.time()
        model2 = KNeighborsClassifier(n_neighbors=k, metric='manhattan')
        model2.fit(x2_train, y2_train)
        model2_accuracy.append(model2.score(x2_test, y2_test))
        t1 = time.time()
        total_time2 = t1 - t0
        model2_time.append(total_time2)
        
            
        t0 = time.time()
        model3 = KNeighborsClassifier(n_neighbors=k, metric='manhattan')
        model3.fit(x3_train, y3_train)
        model3_accuracy.append(model3.score(x3_test, y3_test))
        t1 = time.time()
        total_time3 = t1 - t0
        model3_time.append(total_time3)
        
           
        t0 = time.time()
        model4 = KNeighborsClassifier(n_neighbors=k, metric='manhattan')
        model4.fit(x4_train, y4_train)
        model4_accuracy.append(model4.score(x4_test, y4_test))
        t1 = time.time()
        total_time4 = t1 - t0
        model4_time.append(total_time4)

In [7]:
print("KNN MANHATTAN MODELLING")
print("Model iris mean Accuracies:", mean(model2_accuracy))
print("Model skin mean Accuracies:", mean(model3_accuracy))
print("Model bank mean Accuracies:", mean(model4_accuracy))
print("Model iris mean Times:", mean(model2_time))
print("Model skin mean Times:", mean(model3_time))
print("Model bank mean Times:", mean(model4_time))

KNN MANHATTAN MODELLING
Model iris mean Accuracies: 0.96
Model skin mean Accuracies: 0.9989002693217988
Model bank mean Accuracies: 0.960117302052786
Model iris mean Times: 0.006132054328918457
Model skin mean Times: 2.9271103620529173
Model bank mean Times: 0.238899564743042


In [8]:
from sklearn.naive_bayes import CategoricalNB

#Naive bayes

t0 = time.time()
model1 = CategoricalNB()
model1.fit(x1_train, y1_train)
print('Naive bayes soy performance: ', model1.score(x1_test, y1_test))
t1 = time.time()
total_time1 = t1 - t0
print('Naive bayes soy speed: ',total_time1)
print()
t0 = time.time()
model2 = CategoricalNB()
model2.fit(x2_train, y2_train)
print('Naive bayes iris performance: ', model2.score(x2_test, y2_test))
t1 = time.time()
total_time2 = t1 - t0
print('Naive bayes iris speed: ',total_time2)
print()
t0 = time.time()
model3 = CategoricalNB()
model3.fit(x3_train, y3_train)
print('Naive bayes skin performance: ', model3.score(x3_test, y3_test))
t1 = time.time()
total_time3 = t1 - t0
print('Naive bayes skin speed: ',total_time3)
print()
#t0 = time.time()
#model4 = CategoricalNB()
#model4.fit(x4_train, y4_train)
#print('Naive bayes bank performance: ', model4.score(x4_test, y4_test))
#t1 = time.time()
#total_time4 = t1 - t0
#print('Naive bayes bank speed: ',total_time4)
print("NOTE: NB classification does not work due to the presence of negative values in the BANK dataset")

Naive bayes soy performance:  1.0
Naive bayes soy speed:  0.009855031967163086

Naive bayes iris performance:  0.9333333333333333
Naive bayes iris speed:  0.004797697067260742

Naive bayes skin performance:  0.9447074185913654
Naive bayes skin speed:  0.0345311164855957

NOTE: NB classification does not work due to the presence of negative values in the BANK dataset


In [9]:
from sklearn import tree

#decision trees with varying depth

model1_accuracy = []
model2_accuracy = []
model3_accuracy = []
model4_accuracy = []
model1_time = []
model2_time = []
model3_time = []
model4_time = []
for depth in range(1,30):
        
        t0 = time.time()
        d_tree = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = depth)
        m = d_tree.fit(x1_train, y1_train)
        model1_accuracy.append(m.score(x1_test, y1_test))
        t1 = time.time()
        total_time1 = t1 - t0
        model1_time.append(total_time1)
        
        t0 = time.time()
        d_tree = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = depth)
        m = d_tree.fit(x2_train, y2_train)
        model2_accuracy.append(m.score(x2_test, y2_test))
        t1 = time.time()
        total_time2 = t1 - t0
        model2_time.append(total_time2)
            
        t0 = time.time()
        d_tree = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = depth)
        m = d_tree.fit(x3_train, y3_train)
        model3_accuracy.append(m.score(x3_test, y3_test))
        t1 = time.time()
        total_time3 = t1 - t0
        model3_time.append(total_time3)
        
           
        t0 = time.time()
        d_tree = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = depth)
        m = d_tree.fit(x4_train, y4_train)
        model4_accuracy.append(m.score(x4_test, y4_test))
        t1 = time.time()
        total_time4 = t1 - t0
        model4_time.append(total_time4)

In [10]:
print("TREE CLASSIFIER MODELLING")
print("Model soy mean Accuracies:", mean(model1_accuracy))
print("Model iris mean Accuracies:", mean(model2_accuracy))
print("Model skin mean Accuracies:", mean(model3_accuracy))
print("Model bank mean Accuracies:", mean(model4_accuracy))
print("Model soy mean Times:", mean(model1_time))
print("Model iris mean Times:", mean(model2_time))
print("Model skin mean Times:", mean(model3_time))
print("Model bank mean Times:", mean(model4_time))

TREE CLASSIFIER MODELLING
Model soy mean Accuracies: 0.9758620689655173
Model iris mean Accuracies: 0.9390804597701149
Model skin mean Accuracies: 0.9886804638976522
Model bank mean Accuracies: 0.9471129537870364
Model soy mean Times: 0.0031600409540636786
Model iris mean Times: 0.0027048587799072266
Model skin mean Times: 0.16841128776813374
Model bank mean Times: 0.3577179744325835


In [12]:
import sklearn.ensemble as e

#random forests with varying depth

model1_accuracy = []
model2_accuracy = []
model3_accuracy = []
model4_accuracy = []
model1_time = []
model2_time = []
model3_time = []
model4_time = []
for trees in range(1,101, 10):
        
        t0 = time.time()
        forest = e.RandomForestClassifier(criterion = 'entropy', n_estimators = trees)
        m = forest.fit(x1_train, y1_train)
        model1_accuracy.append(m.score(x1_test, y1_test))
        t1 = time.time()
        total_time1 = t1 - t0
        model1_time.append(total_time1)
        
        t0 = time.time()
        forest = e.RandomForestClassifier(criterion = 'entropy', n_estimators = trees)
        m = forest.fit(x2_train, y2_train)
        model2_accuracy.append(m.score(x2_test, y2_test))
        t1 = time.time()
        total_time2 = t1 - t0
        model2_time.append(total_time2)
            
        t0 = time.time()
        forest = e.RandomForestClassifier(criterion = 'entropy', n_estimators = trees)
        m = forest.fit(x3_train, y3_train)
        model3_accuracy.append(m.score(x3_test, y3_test))
        t1 = time.time()
        total_time3 = t1 - t0
        model3_time.append(total_time3)
        
           
        t0 = time.time()
        forest = e.RandomForestClassifier(criterion = 'entropy', n_estimators = trees)
        m = forest.fit(x4_train, y4_train)
        model4_accuracy.append(m.score(x4_test, y4_test))
        t1 = time.time()
        total_time4 = t1 - t0
        model4_time.append(total_time4)

In [13]:
print("RANDOM FOREST CLASSIFIER MODELLING")
print("Model soy mean Accuracies:", mean(model1_accuracy))
print("Model iris mean Accuracies:", mean(model2_accuracy))
print("Model skin mean Accuracies:", mean(model3_accuracy))
print("Model bank mean Accuracies:", mean(model4_accuracy))
print("Model soy mean Times:", mean(model1_time))
print("Model iris mean Times:", mean(model2_time))
print("Model skin mean Times:", mean(model3_time))
print("Model bank mean Times:", mean(model4_time))

RANDOM FOREST CLASSIFIER MODELLING
Model soy mean Accuracies: 1.0
Model iris mean Accuracies: 0.9333333333333333
Model skin mean Accuracies: 0.9995103240022851
Model bank mean Accuracies: 0.968475073313783
Model soy mean Times: 0.04136018753051758
Model iris mean Times: 0.042569947242736814
Model skin mean Times: 2.361792731285095
Model bank mean Times: 0.9064349174499512


In [14]:
#SVM LINEAR
from sklearn import svm 
from sklearn.svm import LinearSVC

t0 = time.time()
model1 = LinearSVC()
model1.fit(x1_train, y1_train)
print('LINEAR SVM soy performance: ', model1.score(x1_test, y1_test))
t1 = time.time()
total_time1 = t1 - t0
print('LINEAR SVM soy speed: ',total_time1)
print()
t0 = time.time()
model2 =  LinearSVC()
model2.fit(x2_train, y2_train)
print('LINEAR SVM iris performance: ', model2.score(x2_test, y2_test))
t1 = time.time()
total_time2 = t1 - t0
print('LINEAR SVM iris speed: ',total_time2)
print()
t0 = time.time()
model3 =  LinearSVC(max_iter = 1000)
model3.fit(x3_train, y3_train)
print('LINEAR SVM skin performance: ', model3.score(x3_test, y3_test))
t1 = time.time()
total_time3 = t1 - t0
print('LINEAR SVM skin speed: ',total_time3)
print()
t0 = time.time()
model4 =  LinearSVC(max_iter = 1000)
model4.fit(x4_train, y4_train)
print('LINEAR SVM bank performance: ', model4.score(x4_test, y4_test))
t1 = time.time()
total_time4 = t1 - t0
print('LINEAR SVM bank speed: ',total_time4)
print()


LINEAR SVM soy performance:  1.0
LINEAR SVM soy speed:  0.004519224166870117

LINEAR SVM iris performance:  0.8666666666666667
LINEAR SVM iris speed:  0.004996776580810547



C:\Users\Henry Ramstad\anaconda3\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\Henry Ramstad\anaconda3\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


LINEAR SVM skin performance:  0.8698890067738513
LINEAR SVM skin speed:  10.197792768478394

LINEAR SVM bank performance:  0.968475073313783
LINEAR SVM bank speed:  0.6820039749145508



C:\Users\Henry Ramstad\anaconda3\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [18]:
#SVM RBF
 

t0 = time.time()
model1 = svm.SVC()
model1.fit(x1_train, y1_train)
print('SVM soy performance: ', model1.score(x1_test, y1_test))
t1 = time.time()
total_time1 = t1 - t0
print('SVM soy speed: ',total_time1)
print()
t0 = time.time()
model2 =  svm.SVC()
model2.fit(x2_train, y2_train)
print('SVM iris performance: ', model2.score(x2_test, y2_test))
t1 = time.time()
total_time2 = t1 - t0
print('SVM iris speed: ',total_time2)
print()
t0 = time.time()
model3 =  svm.SVC()
model3.fit(x3_train, y3_train)
print('SVM skin performance: ', model3.score(x3_test, y3_test))
t1 = time.time()
total_time3 = t1 - t0
print('SVM skin speed: ',total_time3)
print()
t0 = time.time()
model4 =  svm.SVC()
model4.fit(x4_train, y4_train)
print('SVM bank performance: ', model4.score(x4_test, y4_test))
t1 = time.time()
total_time4 = t1 - t0
print('SVM bank speed: ',total_time4)
print()



SVM soy performance:  0.6
SVM soy speed:  0.0035092830657958984

SVM iris performance:  0.9
SVM iris speed:  0.0020093917846679688

SVM skin performance:  0.9984493593405697
SVM skin speed:  18.957127809524536

SVM bank performance:  0.9706744868035191
SVM bank speed:  0.3119387626647949



ANALYSIS:

for purposes of viewing I will copy and paste the major results of the analysis to compare

NAIVE\
Naive soy model accuracy:  0.4\
Naive iris model accuracy:  0.0\
Naive skin model accuracy:  0.791581653472619\
Naive bank model accuracy:  0.9736070381231672

LOGISTIC\
logistic soy performance:  1.0\
logistic soy speed:  0.010515928268432617

logistic iris performance:  0.9666666666666667\
logistic iris speed:  0.012507200241088867

logistic skin performance:  0.9194687015424794\
logistic skin speed:  0.2614462375640869

logistic bank performance:  0.968475073313783\
logistic bank speed:  0.23293566703796387

KNN EUCLIDEAN MODELLING\
KNN NOT POSSIBLE FOR SOY DATASET*

Model iris mean Accuracies: 0.9933333333333334\
Model iris mean Times: 0.004015636444091797

Model skin mean Accuracies: 0.9992695666367422\
Model skin mean Times: 0.9586779356002808

Model bank mean Accuracies: 0.9730938416422287\
Model bank mean Times: 0.06626036167144775

KNN MANHATTAN MODELLING\
KNN NOT POSSIBLE FOR SOY DATASET*

Model iris mean Accuracies: 0.9800000000000001\
Model iris mean Times: 0.0035050153732299806

Model skin mean Accuracies: 0.999212437770342\
Model skin mean Times: 0.9401741981506347

Model bank mean Accuracies: 0.973607038123167\
Model bank mean Times: 0.09838576316833496

NAIVE BAYES\
Naive bayes soy performance:  1.0\
Naive bayes soy speed:  0.006998777389526367

Naive bayes iris performance:  0.9666666666666667\
Naive bayes iris speed:  0.0030012130737304688

Naive bayes skin performance:  0.9470537827470823\
Naive bayes skin speed:  0.03899717330932617

NOTE: NB classification does not work due to the presence of negative values in the BANK dataset

TREE CLASSIFIER MODELLING\
Model soy mean Accuracies: 0.9827586206896551\
Model soy mean Times: 0.0020212305003199085

Model iris mean Accuracies: 0.954022988505747\
Model iris mean Times: 0.001750946044921875

Model skin mean Accuracies: 0.98875785521913\
Model skin mean Times: 0.1035644350380733

Model bank mean Accuracies: 0.9602083122661543\
Model bank mean Times: 0.1991223220167489

RANDOM FOREST CLASSIFIER MODELLING\
Model soy mean Accuracies: 1.0\
Model soy mean Times: 0.0415095329284668

Model iris mean Accuracies: 0.9533333333333334\
Model iris mean Times: 0.04161231517791748

Model skin mean Accuracies: 0.9995184852689137\
Model skin mean Times: 2.3844290018081664

Model bank mean Accuracies: 0.9733137829912024\
Model bank mean Times: 0.942660140991211




LINEAR SVM\
LINEAR SVM soy performance:  1.0\
LINEAR SVM soy speed:  0.004519224166870117

LINEAR SVM iris performance:  0.8666666666666667\
LINEAR SVM iris speed:  0.004996776580810547

LINEAR SVM skin performance:  0.8698890067738513\
LINEAR SVM skin speed:  10.197792768478394

LINEAR SVM bank performance:  0.968475073313783\
LINEAR SVM bank speed:  0.6820039749145508





RBF SVM\
SVM soy performance:  0.6\
SVM soy speed:  0.0035092830657958984

SVM iris performance:  0.9\
SVM iris speed:  0.0020093917846679688

SVM skin performance:  0.9984493593405697\
SVM skin speed:  18.957127809524536

SVM bank performance:  0.9706744868035191\
SVM bank speed:  0.3119387626647949


NOTE: The data above comes from a singular run (required multipled kernel restarts) which is to explain why the values in cell is different than analysis (different splits).

Key points of analysis:
1. KNN failed for the soybean dataset due to its size
2. NB failed for the bank dataset given its contents contained negative values.
3. SVM for dataset 3 took significant times and had iteration and hang time problems which made it difficult to even evaluate.

soybean:
1. Logistic, Naive bayes, and random forest achieved perfect scores with logistic and NB being significantly faster than random forest

iris:
1. KNN, Logistic, and Naive bayes had the best performance with KNN and NB being the fastest

skin:
1. KNN was by far the best accuracy and fastest with random forests following closely behind.

bank:
1. Every model had similar accuracy for the tested and even the naive approach had a similar result to the models indicating that perhaps the data structure meant the models could not learn much from the underlying function. Overall based on the equal accuracies naive (not naive bayes) and KNN would be fastest.

I think the results above indicate that size (tall versus wide) as well as the type of data contained in the dataset play a large role in what model to select. Additionally, the conditioning of the data in general seems incredibly important or a result like the bank dataset might occur (naive model beating or equalling intensive models like random forest).